astropy/utils/metadata.py

```diff
@@ -23,41 +24,135 @@ class MergeConflictWarning(AstropyWarning):
     pass
 
 
-def take_left(left, right):
-    return left
+MERGE_STRATEGY_CLASSES = []
 
+def common_dtype(arrs):
+    """
+    Use numpy to find the common dtype for a list of ndarrays.
+
+    Only allow arrays within the following fundamental numpy data types:
+    np.bool_, np.object_, np.number, np.character, np.void
+
+    Parameters
+    ----------
+    arrs: list of ndarray objects
+
+    """
+    def dtype(arr):
+        return getattr(arr, 'dtype', np.dtype('O'))
+
+    np_types = (np.bool_, np.object_, np.number, np.character, np.void)
+    uniq_types = set(tuple(issubclass(dtype(arr).type, np_type) for np_type in np_types)
+                     for arr in arrs)
+    if len(uniq_types) > 1:
+        # Embed into the exception the actual list of incompatible types.
+        incompat_types = [dtype(arr).name for arr in arrs]
+        tme = MergeConflictError('Arrays have incompatible types {0}'
+                                 .format(incompat_types))
+        tme._incompat_types = incompat_types
+        raise tme
+
+    arrs = [np.empty(1, dtype=dtype(arr)) for arr in arrs]
+
+    # For string-type arrays need to explicitly fill in non-zero
+    # values or the final arr_common = .. step is unpredictable.
+    for arr in arrs:
+        if arr.dtype.kind in ('S', 'U'):
+            arr[0] = '0' * arr.itemsize
 
-def take_right(left, right):
-    return left
+    arr_common = np.array([arr[0] for arr in arrs])
+    return arr_common.dtype.str
 
 
-def concat(left, right):
+class MergeStrategyMeta(type):
+    """
+    Metaclass that registers MergeStrategy subclasses into the
+    MERGE_STRATEGY_CLASSES registry.
     """
-    Concatenate the ``left`` and ``right`` objects
 
-    Currently only lists and tuples are allowed, but this could change.
+    def __new__(mcls, name, bases, members):
+        cls = super(MergeStrategyMeta, mcls).__new__(mcls, name, bases, members)
+
+        # Register merging class (except for base MergeStrategy class)
+        if 'types' in members:
+            types = members['types']
+            if isinstance(types, tuple):
+                types = [types]
+            for left, right in reversed(types):
+                MERGE_STRATEGY_CLASSES.insert(0, (left, right, cls))
+
+        return cls
+
+@six.add_metaclass(MergeStrategyMeta)
+class MergeStrategy(object):
+    # Set to False to disable applying this merge strategy
+    enabled = True
+
+    @classmethod
+    def merge_with_catch(cls, left, right):
+        """
+        Call the ``merge()`` method, but catch exceptions and re-raise
+
+        Parameters
+        ----------
+        left: object
+            Left object to be merged
+        right: object
+            Right object to be merged
+
+        Returns
+        -------
+        out: object
+            Merged left and right object
+        """
+        try:
+            return cls.merge(left, right)
+        except Exception as err:
+            raise MergeConflictError(err)
```

Comment:
I'm putting on my "Overengineering" hat here, so take with a grain of salt.  But the idea of catching every exception raised by cls.merge and re-raising it as a `MergeConflictError` seems like an implementation detail that doesn't require the API to be designed around it.

Since it seems `merge` is never used directly--only `merge_with_catch`--and since there's already a metaclass for `MergeStrategy` we can automatically wrap all `merge` methods to raise `MergeConflictError` through the metaclass:

```python
from ..utils import wraps
def __new__(mcls, name, bases, members):
    if 'merge' in members and isinstance(members['merge'], classmethod):
        orig_merge = members['merge'].__func__
        @wraps(orig_merge)
        def merge(cls, left, right):
            try:
                return orig_merge(cls, left, right)
            except Exception as err:
                raise MergeConflictError(err)

        members['merge'] = classmethod(merge)
```

(untested, but that's the basic idea).  The modeling package uses this idea a lot so that implementers don't have to worry about invariants expected by the modeling framework.  Of course, your existing approach accomplishes the same with less metaprogramming.  I don't have a strong feeling about which approach to use, though for your approach I might make `_merge_with_catch` internal.

The metaprogramming approach does have the advantage that `MyMergeStrategy.merge` always raises a `MergeConflictError` when used directly, by "magic".